overlord/snapstate/backend/kernel_modules_components.go

```diff
@@ -0,0 +1,217 @@
+	hasModules := osutil.IsDirectory(filepath.Join(mountDir, "modules", kernelVersion))
+	hasFirmware := osutil.IsDirectory(filepath.Join(mountDir, "firmware"))
+	return hasModules, hasFirmware
+}
+
+func componentMountPoint(componentName, kernelVersion string) string {
+	return filepath.Join("/run/mnt/kernel-modules/", kernelVersion, componentName)
+}
+
+func modulesMountPoint(componentName, kernelVersion string) string {
+	return filepath.Join("/usr/lib/modules", kernelVersion, "updates", componentName)
+}
+
+// SetupKernelModulesComponent creates and starts mount units for
+// kernel-modules components.
+func (b Backend) SetupKernelModulesComponent(cpi snap.ContainerPlaceInfo, cref naming.ComponentRef, kernelVersion string, meter progress.Meter) (err error) {
+	var sysd systemd.Systemd
+	if b.preseed {
+		sysd = systemd.NewEmulationMode(dirs.GlobalRootDir)
+	} else {
+		sysd = systemd.New(systemd.SystemMode, meter)
+	}
+
+	// Restore state if something goes wrong
+	var cleanOnFailure kernelModulesCleanupParts
+	defer func() {
+		if err == nil {
+			return
+		}
+		if err := cleanupKernelModulesSetup(&cleanOnFailure, kernelVersion, meter); err != nil {
+			logger.Noticef("while cleaning up a failed kernel-modules set-up: %v", err)
+		}
+	}()
+
+	// Check that the kernel-modules component really has
+	// something we must mount on early boot.
+	hasModules, hasFirmware := checkKernelModulesCompContent(cpi.MountDir(), kernelVersion)
+	if !hasModules && !hasFirmware {
+		return &NoKernelDriversError{cref: cref, kernelVersion: kernelVersion}
+	}
+
+	// Mount the component itself (we need it early, so the mount in /snap cannot
+	// be used).
+	componentMount := componentMountPoint(cref.ComponentName, kernelVersion)
+	addMountUnitOptions := &systemd.MountUnitOptions{
+		MountUnitType: systemd.BeforeDriversLoadMountUnit,
+		Lifetime:      systemd.Persistent,
+		Description:   fmt.Sprintf("Mount unit for kernel-modules component %s", cref),
+		What:          cpi.MountFile(),
+		Where:         componentMount,
+		Fstype:        "squashfs",
+		Options:       []string{"nodev,ro,x-gdu.hide,x-gvfs-hide"},
+	}
+	_, err = sysd.EnsureMountUnitFileWithOptions(addMountUnitOptions)
+	if err != nil {
+		return fmt.Errorf("cannot create mount in %q: %w", componentMount, err)
+	}
+	cleanOnFailure.compMountDir = componentMount
+
+	if hasModules {
+		// systemd automatically works out dependencies on the "what"
+		// path too so this mount happens after the component one.
+		modulesDir := modulesMountPoint(cref.ComponentName, kernelVersion)
+		addMountUnitOptions = &systemd.MountUnitOptions{
+			MountUnitType: systemd.BeforeDriversLoadMountUnit,
+			Lifetime:      systemd.Persistent,
+			Description:   fmt.Sprintf("Mount unit for modules from %s", cref.String()),
+			What:          filepath.Join(componentMount, "modules", kernelVersion),
+			Where:         modulesDir,
+			Fstype:        "none",
+			Options:       []string{"bind"},
```

Comment:
could we record somewhere a discussion of why using a bind mount and not a symlink for this?